proposals/0000-first-class-modules.rst

```diff
@@ -0,0 +1,612 @@
+  - See `Termination`_ below for why this isn’t a problem for this
+    proposal.
+
+Ultimately, I think this proposal is just *simpler*, in that it makes
+modules more like everything else in the module system, which then
+allows us to solve our problems with the namespacing tools we already
+have for other bindings.
+
+There is also an earlier proposal “Structured module exports/imports” [ghc205]_
+I won't respond to it in detail, but it takes the same approach of continuing to
+focus on qualified names rather than giving modules an identity.
+
+Unresolved questions
+--------------------
+
+What should happen if multiple modules are defined with the same name?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+[ghc283]_ effectively proposes that multiple modules with the same name should be
+*merged*. I would like to say that multiple modules with the same name
+should be *ambiguous*, just like all other names.
+
+I think the argument for merging is:
+
+- We have some existing cases where this happens (``import as``), which we need to support.
+- It might be convenient to “mix in” additional names into an existing module.
+
+The argument for ambiguity is:
+
+- Ambiguity is consistent with the way names work for everything else, so makes the system work.
+- It is not very inconvenient in practice
+
+  - Anecdotal evidence: I wrote the Semmle QL module system this way, and I never had anybody ask
+    for module merging.
+  - It is easily to “manually” merge modules if you have local module definitions.
+
+If we thought ambiguity was more consistent with "modules as entities", we could add a language
+extension `AmbiguousAs` which makes such `import as` statements produce ambiguity errors instead.
+
+Do we need any tricky rules for typeclasses/type family instances?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Consider:
+
+::
+
+   module M where
+       instance Ord T where …
+       f :: T -> T -> Bool
+       f = compare
+
+   g = M.f
+
+That is, we can use ``f`` without importing its defining module, and
+hence the typeclass instance is not in scope. Possibly this is fine, but
+I’d like someone who knows more about how typeclasses work to think
+about this with me!
+
+An alternative would be for typeclass instances to be associated with a
+file module rather than any module. So if you import a module which
+contains an instance inside a nested module, then that instance is used
+during resolution even if it is not strictly in scope. But this would be
+somewhat unhygienic.
+
+Can we make dotted names less special?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+File modules declared with dots in them are awkward. The dots *look*
+like the qualifier we use in qualified names, but they behaves
+differently: there is a module called ``Data.Set``, but there is no
+module ``Data`` which contains a module ``Set``.
+
+We could change the system so there *was* such a ``Data`` module. [#]_
```

Comment:
`containers` defines `Data.Set` and `Data.Map`.
`unordered-containers` defines `Data.HashSet` and `Data.HashMap.`

How would these `Data` modules get merged when I depend on both packages?

Reply:
We could merge them just as we do for the "multiple import as" case. In particular, we don't have to have a rule that the resulting modules must have no ambiguous names. We would need something like `PackageImports` to resolve that (as we do today).

Reply:
I've added something to that effect. I think it would actually play nicely with `PackageImports`.